Extend DoubleQDQPairsRemover to handle sequences that end in duplicate DQ nodes

adrianlizarraga opened the pull request:
### Description
Extend the DoubleQDQPairsRemover optimizer to also handle sequences that end in duplicate DQ nodes.

For example, the following sequence:
```
 Q1 --> DQ1 --> Q2 --+--> DQ2
                     |
                     +--> DQ2'
```
Is now simplified to:
```
 Q1 ---+--> DQ2
       |
       +--> DQ2'
```


### Motivation and Context
The EnsureUniqueDQNodeUnits pass may add duplicate DQ nodes to ensure valid QDQ node units. The DoubleQDQPairsRemover should still be able to remove unnecessary QDQ ops if the target sequence ends in duplicate DQ nodes.

adrianlizarraga commented on onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:
```diff
@@ -121,75 +178,80 @@ static bool IsReducibleDoubleQDQSequence(Graph& graph, NodeIndex& q1_index, Node
+  if (q2 == nullptr ||
+      q2->OpType() != "QuantizeLinear" ||
+      graph.NodeProducesGraphOutput(*q2) ||
+      !GetQNodeZeroPointType(graph, *q2, quant_type_2) ||
```
This check is new. Need to make sure that both (Q -> DQ) use the same quantization type. This reduction is invalid otherwise.

yuslepukhin commented on onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:
```diff
@@ -121,75 +178,80 @@ static bool IsReducibleDoubleQDQSequence(Graph& graph, NodeIndex& q1_index, Node
-  const ONNX_NAMESPACE::TensorProto* dq1_zp_tensor_proto = graph.GetConstantInitializer(
-      dq1_input_defs[QDQ::InputIndex::ZERO_POINT_ID]->Name(), true);
+  // All of q2's children should be DQ nodes with zero-point and scale values equal to those of q2.
+  std::vector<Node*> dq2_nodes;
```
InlinedVector everywhere. For 64-bit ptrs it can fit up to 6 wo memory allocations

adrianlizarraga replied:
Updated

adrianlizarraga commented on onnxruntime/test/optimizer/graph_transform_test_builder.h:
```diff
@@ -195,21 +196,20 @@ class ModelTestBuilder {
+  /// <param name="elem_type">ONNX tensor element data type</param>
+  /// <param name="raw_data">Raw data bytes</param>
+  /// <returns>NodeArg pointer</returns>
+  NodeArg* MakeInitializer(gsl::span<const int64_t> shape, ONNX_NAMESPACE::TensorProto_DataType elem_type,
```
I tried doing this, but this change would require modifying a large number of tests across various files. The main issue is that passing an initializer list, such as `{1, 2, 3}`, as an argument for a `gsl::span` parameter fails to compile. Would need to modify all call sites to pass in `AsSpan(std::array{1, 2, 3})` instead.

Although I think this would be a good cleanup task, I would prefer to keep this PR more focused.

edgchen1 commented on onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:
```diff
@@ -121,75 +178,80 @@ static bool IsReducibleDoubleQDQSequence(Graph& graph, NodeIndex& q1_index, Node
-  Node* dq2 = graph.GetNode(dq2_index);
-  if (dq2 == nullptr ||
-      dq2->OpType() != "DequantizeLinear") {
+  int64_t quant_type = ONNX_NAMESPACE::TensorProto_DataType_UNDEFINED;
```
nit: naming suggestion - q1_quant_type and q2_quant_type

edgchen1 commented on onnxruntime/test/optimizer/graph_transform_test_builder.cc:
```diff
@@ -20,6 +21,90 @@
+  switch (type) {
+    case ONNX_NAMESPACE::TensorProto_DataType_INT8: {
+      int8_t val = static_cast<int8_t>(zero_point);
+      auto span = ReinterpretAsSpan<const std::byte, const int8_t>(gsl::make_span(&val, 1));
```
would using gsl::span::as_bytes be simpler?

adrianlizarraga replied:
Yes, good call. I updated this.

adrianlizarraga commented on onnxruntime/test/optimizer/graph_transform_test_builder.h:
```diff
@@ -6,6 +6,7 @@
 #include <type_traits>
 #include <vector>
 
+#include "core/common/gsl.h"
```
Updated

adrianlizarraga commented on onnxruntime/test/optimizer/graph_transform_test_builder.cc:
```diff
@@ -20,6 +21,90 @@
+    case ONNX_NAMESPACE::TensorProto_DataType_INT8: {
+      int8_t val = static_cast<int8_t>(zero_point);
+      auto span = ReinterpretAsSpan<const std::byte, const int8_t>(gsl::make_span(&val, 1));
+      return std::vector<std::byte>(span.begin(), span.end());
```
Updated

adrianlizarraga commented on onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:
```diff
@@ -81,38 +112,64 @@ static bool FindNewZeroPointAndScale(const Graph& graph, const Node& node1, cons
-static bool RecomputeOuterQDQZeroPointAndScale(Graph& graph, Node& q1, const Node& dq1, const Node& q2, Node& dq2) {
-  bool skip_reset = false;
+static bool RecomputeOuterQDQZeroPointAndScale(Graph& graph, Node& q1, const Node& dq1, const Node& q2,
+                                               gsl::span<Node*> dq2s) {
```
Updated

adrianlizarraga commented on onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:
```diff
@@ -2,13 +2,44 @@
+    // If a zero_point input is absent, get the type from the "output_dtype" attribute or default to uint8.
+    // The "output_dtype" attribute was added in ONNX opset 21.
+    const auto* attr = graph_utils::GetNodeAttribute(q_node, "output_dtype");
+    zp_data_type = attr != nullptr ? attr->i() : static_cast<int64_t>(ONNX_NAMESPACE::TensorProto_DataType_UINT8);
```
Thanks. I missed this. Updated.

HectorSVC reviewed (approved):
:shipit:

adrianlizarraga commented:
@yuslepukhin I addressed your comments. Could you please take another look? Merging is blocked on Github until you approve that your requested changes have been addressed.
